semantic/constraints.go

```diff
@@ -338,11 +341,37 @@ func (v ConstraintGenerator) typeof(n Node) (PolyType, error) {
 			properties[field.Key.Key()] = t
 			upper = append(upper, field.Key.Key())
 		}
+		var with *Tvar
+		if n.With != nil {
+			t, err := v.lookup(n.With)
+			if err != nil {
+				return nil, err
+			}
+			tv, ok := t.(Tvar)
+			if !ok {
+				return nil, errors.New("object 'with' identifier must be a type variable")
+			}
+			for _, k := range v.cs.kindConst[tv] {
+				obj, ok := k.(ObjectKind)
+				if !ok {
+					return nil, errors.New("object 'with' identifier must be have only object kind constraints")
+				}
+				if obj.upper.isAllLabels() {
+					continue
+				}
+				for _, p := range obj.upper {
+					properties[p] = obj.properties[p]
+				}
+				upper = upper.union(obj.upper)
+			}
+		}
 		v.cs.AddKindConst(nodeVar, ObjectKind{
+			with:       with,
```

Comment:
This is I think where it would be useful for `with` to be an operator. The constraint phase for object expressions does not need to change. Instead we could add a case for `WithExpression`, and all that would need to be done I believe is to create a kind constraint for the new object expression. And this would be achieved by just taking the union of the lower labels and the upper labels.

Something like this:
```go
case *WithExpression:
    ...
    leftLowerLabels := ...
    rightLowerLabels := ...

    leftUpperLabels := ...
    rightUpperLabels := ...

    v.cs.AddKindConst(nodeVar, ObjectKind{
        properties: properties,
        lower: leftLowerLables.union(rightLowerLabels),
        upper: leftUpperLabels.union(rightUpperLabels),
    }
```

Reply:
this looks so much easier to maintain.  the `with` as part of the definition gets messy really fast.

Reply:
I like it.

What happens if the field already exists?  I guess it would make sense for the RHS of `with` override what's already there.  E.g., what should these do?
```
// (1)
{foo: 1.0} with {foo: 2.0}
// (2)
{foo: 1.0} with {foo: "two dot oh"}
```

Reply:
So it turns out it's not quite as easy as I'm making it out to be. While I definitely prefer the syntax, my suggestion above won't work with functions that take records as input. So this won't type check:

    f = (r) => r with {foo: "bar"}

And we sort of need this to type check

Now we can work around this for row functions since row functions in transformations like map and filter aren't really type checked as strictly as functions defined outside of such transformations. But I'd rather have all functions, including row functions, be treated the same way by the type system.

So how to support the `with` syntax is still an open question for our type system right now.

To your question @wolffcm it makes sense to override to me as well.